enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+#### Bootstrap changes
+
+For the initial transition from in-tree to the out-of-tree, `CCM` pod will be created as static pod on the bootstrap node, to ensure swift removal of the `node.cloudprovider.kubernetes.io/uninitialized: true` taint from any newly created `Nodes`. Later stages, including cluster upgrades will be managed by an operator, which will ensure stability of the configuration, and will run a `CCM` in a `Deployment`. Initial usage of the static pod is justified by the need to run `CCM` pod before the `kube-apiserver`, `etcd` and the `kube-scheduler` will get healthy, so the functionality does not rely on them.
+
+`CCM` operator will provide initial manifests that allow to deploy `CCM` on the bootstrap machine with `bootkube.sh` [script](https://github.com/openshift/installer/blob/master/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template).
+
+At the initial stage of `CCM` installation installer creates a config map with `cloud.conf` key that contains configuration of `CCM` in `ini` format. The contents of `cloud.conf` are static and generated by the installer:
+
+Example for OpenStack:
+
+```txt
+[Global]
+secret-name = openstack-credentials
+secret-namespace = kube-system
+```
+
+Resources deployed here will be destroyed with the bootstrap stage cleanup.
+
+#### Operator resource management strategy
+
+Every cloud manages it's configuration differently, but shares a common overall structure.
+
+1. `cloud-controller-manager` and related resources are deployed under a single namespace.
+2. Each pod (`CCM`, `cloud-node-manager`, etc.) will be managed by a `Deployment`.
+3. The pods are expected to run on the control-plane `Nodes` using `hostNetwork`.
+
+Operator will manage:
+
+- Dynamic provider resources, such as `Deployments`.
+- Operator will populate, create and manage the cloud-specific set of `Deployments`, based on the selected `platform` in the `Infrastructure` resource.
+- `Deployments`, provisioned by the operator will run the pods on the `control-plane` nodes, and will tolerate `cloudprovider.kubernetes.io/uninitialized` and `node-role.kubernetes.io/master` taints. They will run at least `2` replicas of the `CCM` pods with anti-affinity settings to prevent scheduling on the same machine and allow some fault-tolerance.
+- The `--cloud-provider` and the `--cloud-config` arguments will be populated using existing `library-go` based implementation from the [cluster-kube-controller-manager-operator](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/990ea3f2ace13be6579c00a889842c3f5d3e756a/pkg/operator/configobservation/configobservercontroller/observe_config_controller.go#L82-L85) to share a common approach with the bootstrap specific static-pod configuration.
```

Comment:
I'm not quite sure what you mean with this statement. You've already described KCM-related changes sooner. Most probably you're again going into unnecessary details.